cxx/hirm.hh

```diff
@@ -191,20 +196,22 @@ class Domain {
   }
 };
 
+template <typename DistributionType>
```

Comment:
It would be nice to have a comment here saying "DistributionType should inherit from base.hh::Distribution".

Or you could do a static_assert:

#include <type_traits>

....

class Relation {
  ...
  static_assert(std::is_base_of<Distribution<ValueType>, DType>::value,
                          "DistributionType must inherit from Distribution");
 ...
};

Reply:
`util_io` still needs updating, I'll do that and address the other comments next.